From eee04ad6a243470099bba98cebab689e2dbf7d07 Mon Sep 17 00:00:00 2001
From: Ilie Halip <ilie.halip@gmail.com>
Date: Thu, 5 Mar 2020 13:20:33 +0200
Subject: [PATCH] elfnote: always use the "a" flag for .note.* sections

GNU as allows specifying different flags when constructing sections, for
example when the second occurrence has empty flags. But clang may not be so
permissive:
    $ cat test.s
    .pushsection .note.Linux, "a", @note ;
    .pushsection .note.Linux, "", @note ;
    $ as test.s
    $ clang-11 -cc1as -triple x86_64-pc-linux-gnu test.s
    test.s:2:1: error: changed section flags for .note.Linux

The kernel reaches this case when both ELFNOTE_START and ELFNOTE add data
to the same section, because the former is used exclusively with the "a"
flag and the latter always uses empty flags.

In vmlinux, the combined .notes section is marked as SHF_ALLOC, so use
the "a" flag for all these macros.

Link: https://github.com/ClangBuiltLinux/linux/issues/913
Signed-off-by: Ilie Halip <ilie.halip@gmail.com>
---
 arch/arm64/kernel/vdso/note.S      | 2 +-
 arch/mips/vdso/elf.S               | 2 +-
 arch/nds32/kernel/vdso/note.S      | 2 +-
 arch/s390/kernel/vdso64/note.S     | 2 +-
 arch/sparc/vdso/vdso-note.S        | 2 +-
 arch/sparc/vdso/vdso32/vdso-note.S | 2 +-
 arch/x86/entry/vdso/vdso-note.S    | 2 +-
 arch/x86/entry/vdso/vdso32/note.S  | 4 ++--
 arch/x86/um/vdso/vdso-note.S       | 2 +-
 include/linux/elfnote.h            | 6 +++---
 10 files changed, 13 insertions(+), 13 deletions(-)

diff --git a/arch/arm64/kernel/vdso/note.S b/arch/arm64/kernel/vdso/note.S
index 0ce6ec75a5252..080fe6c61d3a6 100644
--- a/arch/arm64/kernel/vdso/note.S
+++ b/arch/arm64/kernel/vdso/note.S
@@ -13,7 +13,7 @@
 #include <linux/elfnote.h>
 #include <linux/build-salt.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
 
diff --git a/arch/mips/vdso/elf.S b/arch/mips/vdso/elf.S
index a25cb147f1caa..76a543df664d9 100644
--- a/arch/mips/vdso/elf.S
+++ b/arch/mips/vdso/elf.S
@@ -11,7 +11,7 @@
 #include <linux/elfnote.h>
 #include <linux/version.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
 
diff --git a/arch/nds32/kernel/vdso/note.S b/arch/nds32/kernel/vdso/note.S
index 0aeaa19b05f03..fbf9175e31ab2 100644
--- a/arch/nds32/kernel/vdso/note.S
+++ b/arch/nds32/kernel/vdso/note.S
@@ -6,6 +6,6 @@
 #include <linux/version.h>
 #include <linux/elfnote.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
diff --git a/arch/s390/kernel/vdso64/note.S b/arch/s390/kernel/vdso64/note.S
index db19d0680a0af..2eb232beabb73 100644
--- a/arch/s390/kernel/vdso64/note.S
+++ b/arch/s390/kernel/vdso64/note.S
@@ -8,6 +8,6 @@
 #include <linux/version.h>
 #include <linux/elfnote.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
diff --git a/arch/sparc/vdso/vdso-note.S b/arch/sparc/vdso/vdso-note.S
index 79a071e4357e4..d7c90b620c679 100644
--- a/arch/sparc/vdso/vdso-note.S
+++ b/arch/sparc/vdso/vdso-note.S
@@ -7,6 +7,6 @@
 #include <linux/version.h>
 #include <linux/elfnote.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
diff --git a/arch/sparc/vdso/vdso32/vdso-note.S b/arch/sparc/vdso/vdso32/vdso-note.S
index e234983cf0d81..ebfd87316ccee 100644
--- a/arch/sparc/vdso/vdso32/vdso-note.S
+++ b/arch/sparc/vdso/vdso32/vdso-note.S
@@ -7,6 +7,6 @@
 #include <linux/version.h>
 #include <linux/elfnote.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long	LINUX_VERSION_CODE
 ELFNOTE_END
diff --git a/arch/x86/entry/vdso/vdso-note.S b/arch/x86/entry/vdso/vdso-note.S
index 79423170118f7..4798d2f437390 100644
--- a/arch/x86/entry/vdso/vdso-note.S
+++ b/arch/x86/entry/vdso/vdso-note.S
@@ -8,7 +8,7 @@
 #include <linux/version.h>
 #include <linux/elfnote.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
 
diff --git a/arch/x86/entry/vdso/vdso32/note.S b/arch/x86/entry/vdso/vdso32/note.S
index e78047d119f6b..bb218cb84e4e9 100644
--- a/arch/x86/entry/vdso/vdso32/note.S
+++ b/arch/x86/entry/vdso/vdso32/note.S
@@ -11,7 +11,7 @@
 /* Ideally this would use UTS_NAME, but using a quoted string here
    doesn't work. Remember to change this when changing the
    kernel's name. */
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
 
@@ -39,7 +39,7 @@ BUILD_SALT
 
 #include "../../xen/vdso.h"	/* Defines VDSO_NOTE_NONEGSEG_BIT.  */
 
-ELFNOTE_START(GNU, 2, "a")
+ELFNOTE_START(GNU, 2)
 	.long 1			/* ncaps */
 VDSO32_NOTE_MASK:		/* Symbol used by arch/x86/xen/setup.c */
 	.long 0			/* mask */
diff --git a/arch/x86/um/vdso/vdso-note.S b/arch/x86/um/vdso/vdso-note.S
index 79a071e4357e4..d7c90b620c679 100644
--- a/arch/x86/um/vdso/vdso-note.S
+++ b/arch/x86/um/vdso/vdso-note.S
@@ -7,6 +7,6 @@
 #include <linux/version.h>
 #include <linux/elfnote.h>
 
-ELFNOTE_START(Linux, 0, "a")
+ELFNOTE_START(Linux, 0)
 	.long LINUX_VERSION_CODE
 ELFNOTE_END
diff --git a/include/linux/elfnote.h b/include/linux/elfnote.h
index f236f5b931b2a..1f9f0827cf5f4 100644
--- a/include/linux/elfnote.h
+++ b/include/linux/elfnote.h
@@ -39,8 +39,8 @@
  * e.g. ELFNOTE(XYZCo, 42, .asciz, "forty-two")
  *      ELFNOTE(XYZCo, 12, .long, 0xdeadbeef)
  */
-#define ELFNOTE_START(name, type, flags)	\
-.pushsection .note.name, flags,@note	;	\
+#define ELFNOTE_START(name, type)		\
+.pushsection .note.name, "a",@note	;	\
   .balign 4				;	\
   .long 2f - 1f		/* namesz */	;	\
   .long 4484f - 3f	/* descsz */	;	\
@@ -54,7 +54,7 @@
 .popsection				;
 
 #define ELFNOTE(name, type, desc)		\
-	ELFNOTE_START(name, type, "")		\
+	ELFNOTE_START(name, type)		\
 		desc			;	\
 	ELFNOTE_END
 
